pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+
+		// increment nonce
+		streamNonce = streamNonce + 1
+		out.Nonce = strconv.FormatInt(streamNonce, 10)
+		if s.callbacks != nil {
+			s.callbacks.OnStreamDeltaResponse(streamID, resp.GetDeltaRequest(), out)
+		}
+
+		return out.Nonce, str.Send(out)
+	}
+
+	// setState updates the currently known state of resources in the server
+	setState := func(res cache.DeltaResponse, typeURL string) {
+		values.mu.Lock()
+		defer values.mu.Unlock()
+
+		state := values.deltaStreamStates[typeURL]
+		if state.ResourceVersions == nil {
+			state.ResourceVersions = make(map[string]string)
+		}
+		state.ResourceVersions = res.GetNextVersionMap()
+		values.deltaStreamStates[typeURL] = state
+	}
+
+	if s.callbacks != nil {
+		if err := s.callbacks.OnDeltaStreamOpen(str.Context(), streamID, defaultTypeURL); err != nil {
+			return err
+		}
+	}
+
+	// node may only be set on the first discovery request
+	var node = &core.Node{}
+	isWildcard := map[string]bool{}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+			// config watcher can send the requested resources types in any order
+		case resp, more := <-values.deltaEndpoints:
+			if !more {
+				return status.Errorf(codes.Unavailable, "endpoints watch failed")
+			}
+			nonce, err := send(resp, resource.EndpointType)
+			if err != nil {
+				return err
+			}
+			values.deltaEndpointNonce = nonce
+			setState(resp, resource.EndpointType)
+		case resp, more := <-values.deltaClusters:
+			if !more {
+				return status.Errorf(codes.Unavailable, "clusters watch failed")
+			}
+			nonce, err := send(resp, resource.ClusterType)
+			if err != nil {
+				return err
+			}
+			values.deltaClusterNonce = nonce
+			setState(resp, resource.ClusterType)
+		case resp, more := <-values.deltaRoutes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "routes watch failed")
+			}
+			nonce, err := send(resp, resource.RouteType)
+			if err != nil {
+				return err
+			}
+			values.deltaRouteNonce = nonce
+			setState(resp, resource.RouteType)
+		case resp, more := <-values.deltaListeners:
+			if !more {
+				return status.Errorf(codes.Unavailable, "listeners watch failed")
+			}
+			nonce, err := send(resp, resource.ListenerType)
+			if err != nil {
+				return err
+			}
+			values.deltaListenerNonce = nonce
+			setState(resp, resource.ListenerType)
+		case resp, more := <-values.deltaSecrets:
+			if !more {
+				return status.Errorf(codes.Unavailable, "secrets watch failed")
+			}
+			nonce, err := send(resp, resource.SecretType)
+			if err != nil {
+				return err
+			}
+			values.deltaSecretNonce = nonce
+			setState(resp, resource.SecretType)
+		case resp, more := <-values.deltaRuntimes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "runtimes watch failed")
+			}
+			nonce, err := send(resp, resource.RuntimeType)
+			if err != nil {
+				return err
+			}
+			values.deltaRuntimeNonce = nonce
+			setState(resp, resource.RuntimeType)
+		case resp, more := <-values.deltaResponses:
+			if more {
```

Comment:
why does this branch look different?

Reply:
I think that was a sotw change to deal with the mux cache. Honestly this logic makes more sense so I pulled it back into the other branches in this refactor I have going